Unit tests for Evm Binding Objects

fabrobles92 opened the pull request:
Closes #41 

Unit tests for concreteBlockContext  and concreteEVM methods .

fabrobles92 committed:
core: create unit tests for ConcreteBlockContext methods

fabrobles92 commented:
Hello @therealbytes ,
Looking for early feedback, pls let me know if something is not aligned with what you expect.

Thanks in advance :D

therealbytes commented:
The block context binding tests are directionally correct but need some changes:
- No need to create so many separate tests functions that all do almost the same thing. We can put everything inside the same parent function.
- If you use a utility function to create a test block context, return a block context, not a pointer to a block context that you immediately reference every time.

This should be a good start to fix both of these issues:

```golang
func TestConcreteBlockContext(t *testing.T) {
	r := require.New(t)

	var (
		block0Hash = crypto.Keccak256Hash([]byte("test0"))
		block1Hash = crypto.Keccak256Hash([]byte("test1"))
		randomHash = crypto.Keccak256Hash([]byte("random hash"))
	)

	getHash := func(blockNumber uint64) common.Hash {
		switch blockNumber {
		case 0:
			return block0Hash
		case 1:
			return block1Hash
		}
		return common.Hash{}
	}

	blockCtx := BlockContext{
		BlockNumber: big.NewInt(123),
		GetHash:     getHash,
		Time:        456,
		GasLimit:    789,
		Difficulty:  big.NewInt(123456),
		BaseFee:     big.NewInt(456123),
		Coinbase:    common.HexToAddress("0x1234567890abcdef1234567890abcdef1234567890"),
		Random:      &randomHash,
	}

	ccBlockCtx := concreteBlockContext{ctx: &blockCtx}

	t.Run("GetHash", func(t *testing.T) {
		r.Equal(block0Hash, ccBlockCtx.GetHash(0))
		r.Equal(block1Hash, ccBlockCtx.GetHash(1))
	})
	t.Run("Timestamp", func(t *testing.T) {
		r.Equal(blockCtx.Time, ccBlockCtx.Timestamp())
	})
	t.Run("BlockNumber", func(t *testing.T) {
		r.Equal(blockCtx.BlockNumber.Uint64(), ccBlockCtx.BlockNumber())
	})
	// ...
}
```

Also, please keep the naming consistent and don't change the capitalization of compound words like `Timestamp`. It should be `Timestamp`, not `TimeStamp`. Geth already has some inconsistencies here (e.g., `Blockchain` vs. `BlockChain`) and we don't want to make it worse!

fabrobles92 committed:
-Apply code review comments from bytes
-Unit test for CallStatic
-Unit test for Call
-Unit test for Create
-Unit test for Create2

therealbytes commented:
Just merged #50 to fix the delegate call issue and add an `extCaller` interface. Rather than passing a caller address to `concreteEVM`, we pass a `*Contract`. These are the changes you need to make to your tests:
```golang
var (
	// ...
	self       = common.BytesToAddress([]byte("self"))
	caller     = common.BytesToAddress([]byte("caller"))
	contract   = NewContract(AccountRef(caller), AccountRef(self), new(uint256.Int), 10_000_000)
)

// ...

ccVmenv := concreteEVM{
	evm:      vmenv,
	contract: contract,
}

// ...

t.Run("Create", func(t *testing.T) {
	expectedContractAddr := crypto.CreateAddress(ccVmenv.contract.Address(), statedb.GetNonce(ccVmenv.contract.Address()))
	// ...
})

// ...
```

Once you have that working we can go over some nitpicks in variable naming. We will merge this in and I will then open another issue to repurpose these tests as Concrete Environment tests and use the new `extCaller` interface to make the binding tests simpler.

fabrobles92 committed:
Broken CallDelegate test

fabrobles92 committed:
Merge remote-tracking branch 'origin/main' into Issue#41/test-concrete-bindings

fabrobles92 committed:
Updating tests to use new interface and contract object when creating a concreteEVM

therealbytes commented:
I made some illustrative changes to some of your tests:

```golang
package vm

import (
	"math/big"
	"testing"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/core/rawdb"
	"github.com/ethereum/go-ethereum/core/state"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/params"
	"github.com/holiman/uint256"
	"github.com/stretchr/testify/require"
)

func newTestBlockContext() BlockContext {
	var (
		block0Hash = crypto.Keccak256Hash([]byte("block0"))
		block1Hash = crypto.Keccak256Hash([]byte("block1"))
		randomHash = crypto.Keccak256Hash([]byte("random"))
	)

	getHash := func(blockNumber uint64) common.Hash {
		switch blockNumber {
		case 0:
			return block0Hash
		case 1:
			return block1Hash
		}
		return common.Hash{}
	}

	blockCtx := BlockContext{
		BlockNumber: big.NewInt(0),
		GetHash:     getHash,
		Time:        50,
		GasLimit:    1000,
		Difficulty:  big.NewInt(1234),
		BaseFee:     big.NewInt(5678),
		Coinbase:    common.HexToAddress("0x1234123412341234123412341234123412341234"),
		Random:      &randomHash,
	}

	return blockCtx
}

func TestConcreteBlockContext(t *testing.T) {
	var (
		r          = require.New(t)
		blockCtx   = newTestBlockContext()
		ccBlockCtx = concreteBlockContext{ctx: &blockCtx}
	)
	t.Run("GetHash", func(t *testing.T) {
		r.Equal(blockCtx.GetHash(0), ccBlockCtx.GetHash(0))
		r.Equal(blockCtx.GetHash(1), ccBlockCtx.GetHash(1))
	})
	t.Run("Timestamp", func(t *testing.T) {
		r.Equal(blockCtx.Time, ccBlockCtx.Timestamp())
	})
	t.Run("BlockNumber", func(t *testing.T) {
		r.Equal(blockCtx.BlockNumber.Uint64(), ccBlockCtx.BlockNumber())
	})
	t.Run("GasLimit", func(t *testing.T) {
		r.Equal(blockCtx.GasLimit, ccBlockCtx.GasLimit())
	})
	t.Run("Difficulty", func(t *testing.T) {
		r.Equal(uint256.MustFromBig(blockCtx.Difficulty), ccBlockCtx.Difficulty())
	})
	t.Run("BaseFee", func(t *testing.T) {
		r.Equal(uint256.MustFromBig(blockCtx.BaseFee), ccBlockCtx.BaseFee())
	})
	t.Run("Coinbase", func(t *testing.T) {
		r.Equal(blockCtx.Coinbase, ccBlockCtx.Coinbase())
	})
	t.Run("Random", func(t *testing.T) {
		r.Equal(*blockCtx.Random, ccBlockCtx.Random())
	})
}

func TestEVMCallStatic(t *testing.T) {
	var (
		r            = require.New(t)
		statedb, _   = state.New(types.EmptyRootHash, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
		callerAddr   = common.BytesToAddress([]byte("caller"))
		selfAddr     = common.BytesToAddress([]byte("self"))
		externalAddr = common.BytesToAddress([]byte("external"))
	)

	vm := NewEVM(BlockContext{}, TxContext{}, statedb, params.TestChainConfig, Config{})
	contract := NewContract(AccountRef(callerAddr), AccountRef(selfAddr), new(uint256.Int), 10_000_000)
	ccVm := concreteEVM{evm: vm, contract: contract}

	gas := uint64(10_000)
	slot := common.BytesToHash([]byte("key"))
	value := common.BytesToHash([]byte("value"))

	// Bytecode to load the passed slot address from storage and return it
	statedb.SetCode(externalAddr, hexutil.MustDecode("0x6000355460005260206000F3"))
	statedb.SetState(externalAddr, slot, value)

	ret, remainingGas, err := ccVm.CallStatic(externalAddr, slot.Bytes(), gas)

	r.NoError(err)
	r.Equal(value.Bytes(), ret)
	r.Less(remainingGas, gas)
}

func TestEVMCallDelegate(t *testing.T) {
	var (
		r            = require.New(t)
		statedb, _   = state.New(types.EmptyRootHash, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
		callerAddr   = common.BytesToAddress([]byte("caller"))
		selfAddr     = common.BytesToAddress([]byte("self"))
		externalAddr = common.BytesToAddress([]byte("external"))
	)

	vm := NewEVM(BlockContext{}, TxContext{}, statedb, params.TestChainConfig, Config{})
	contract := NewContract(AccountRef(callerAddr), AccountRef(selfAddr), new(uint256.Int), 10_000_000)
	ccVm := concreteEVM{evm: vm, contract: contract}

	gas := uint64(10_000)
	slot := common.BytesToHash([]byte("key"))
	value := common.BytesToHash([]byte("value"))

	// Bytecode to load the passed slot address from storage and return it
	statedb.SetCode(externalAddr, hexutil.MustDecode("0x6000355460005260206000F3"))
	statedb.SetState(selfAddr, slot, value)

	ret, remainingGas, err := ccVm.CallDelegate(externalAddr, slot.Bytes(), gas)

	r.NoError(err)
	r.Equal(value.Bytes(), ret)
	r.Less(remainingGas, gas)
}
```

Pay close attention to the changes I have made an incorporate them in the remaining tests. Some things to keep in mind:
- Keep things readable, maintainable, and simple. In that order.
- Use accurate, consistent, clear, descriptive names.
- Don't repeat yourself more than necessary (e.g., if your are checking a value equals the expected, and the expected is explicitly set to a non-nil value, there is no need to check the value is not nil as well).
- Test what you are testing, don't test what you are not testing.

If some of the changes I have made don't make sense to you or you think they are a bad idea, please let me know!

fabrobles92 commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	blockCtx := BlockContext{
+		CanTransfer: func(StateDB, common.Address, *uint256.Int) bool { return true },
```
@therealbytes  I leave this one cause is needed for .create() and create2()

fabrobles92 commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,215 @@
+
+	blockCtx := BlockContext{
+		CanTransfer: func(StateDB, common.Address, *uint256.Int) bool { return true },
+		Transfer:    func(StateDB, common.Address, common.Address, *uint256.Int) {},
```
@therealbytes I leave this one cause is needed for .call(), .create() and create2()

fabrobles92 committed:
core:  -DelegateCall test working
-Refactor in all tests
-Contract bytecode that set storage in TestEVMCall

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+	"github.com/stretchr/testify/require"
+)
+
+func newBlockContext() BlockContext {
```
This should be `newTestBlockContext`.
- It is not creating a generic block context but one for testing specifically. This should be reflected in the name.
- If someone wanted to add a `newBlockContext` non-test function to the `vm` package, it would create a name collision.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+		GasLimit:    1000,
+		Difficulty:  big.NewInt(23456146),
+		BaseFee:     big.NewInt(739657255),
+		Coinbase:    common.HexToAddress("0x0854167430392BBc2D15Dd1Cc17e761897AF31C9"),
```
The value of these fields is not important. What is important as that they are easily recognizable as not important. It should be clear for readers that they could hold any other positive value and the test would still pass. `1234` and `5678` are easily recognizable, `739657255` is not.

I forgot to change this before, but BlockNumber should not be 0. It seems more likely for the binding to malfunction in a way that returns 0 than in a way that returns an arbitrary positive number.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+	var (
+		block0Hash = crypto.Keccak256Hash([]byte("block0"))
+		block1Hash = crypto.Keccak256Hash([]byte("block1"))
+	)
```
We can get these values with `blockCtx.GetHash`. There is not need to hard-code them again.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+
+	t.Run("Random", func(t *testing.T) {
+		r.Equal(*blockCtx.Random, ccBlockCtx.Random())
+	})
```
I feel like the `t.Run` calls provide sufficient visual spacing between tests. Adding empty lines spreads them too far apart. Happy to hear an argument in favor of the empty lines, though.

fabrobles92 replied:
Not a particular reason for doing this. I felt more comfortable seeing them with spaces, more a personal reason than a technical one. 

Resolving this without spaces.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+	r.NoError(err)
+	r.Equal(value.Bytes(), ret)
+	r.Less(remainingGas, gas)
+
```
I don't see the point in an empty line between the last statement and the closing bracket. Same for `TestEVMCallDelegate`.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+
+	ret, remainingGas, err := ccVm.Call(externalAddr, value.Bytes(), gas, new(uint256.Int))
+	r.NoError(err)
+	r.Equal(ret, value.Bytes())
```
The first value of an Equal call should be the expected value, the second should be the actual value.

therealbytes commented on core/vm/concrete_bindings_test.go:
I was wrong when I sad `vmenv` and `vmctx` where not consitent with the rest of the codebase. Some other tests use this naming. We shouldn't use it as it is not clear/accurate. The one I proposed for the EVM is not quite consistent either. I think we should go with `evm` for the EVM, and `ccEvm` for the binding. We can keep `blockCtx` for the block context as it is accurate and consistent.

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+	gas := uint64(10_000_000)
+	value := common.BytesToHash([]byte("value"))
+	// Bytecode that performs a storage opcode and return the value passed
+	statedb.SetCode(externalAddr, hexutil.MustDecode("0x6000356000355560006000355460005260206000F3"))
```
I don't like the program and I don't like how it is commented. The storage operation should be influenced by the input and the output should be the output of some operation. Also, from the comment alone it is not clear at all what "performs a storage opcode" means. How about using this code:

```golang
gas := uint64(10_000_000)
slot := common.BytesToHash([]byte("key"))
ogValue := common.BytesToHash([]byte("original"))
newValue := common.BytesToHash([]byte("new"))
input := append(slot.Bytes(), newValue.Bytes()...)

// Replace the value at the passed slot with the passed value and return the original value
statedb.SetCode(externalAddr, hexutil.MustDecode("0x600035546020356000355560005260206000F3"))
statedb.SetState(externalAddr, slot, ogValue)
statedb.AddAddressToAccessList(externalAddr)

ret, remainingGas, err := ccEvm.Call(externalAddr, input, gas, new(uint256.Int))
r.NoError(err)
r.Equal(ogValue.Bytes(), ret)
r.Equal(newValue, statedb.GetState(externalAddr, slot))
r.Less(remainingGas, gas)
```

fabrobles92 replied:
I like it better

therealbytes commented on core/vm/concrete_bindings_test.go:
```diff
@@ -0,0 +1,211 @@
+		r.NoError(err, "Call should not return an error")
+		r.Equal(statedb.GetCode(addr), ret, "State not correct")
+		r.Less(remainingGas, gas, "Gas used should be less than the provided gas")
+	})
```
- Expected value should go before actual value in r.Equal.
- The stored bytecode is checked against `ret` but `ret` isn't checked anywhere. If the binding was malfunctioning in a way that returns and empty byte slice and doesn't set the code, the test would still pass.
- We should check the address is correct (like you were doing before) since, even though we are not testing the EVM itself, we must check the binding is returning the correct value.
- There should be a comment describing what the bytecode does.
- Using a sum for the creation code is not wrong per se but it is not aligned with what a smart contract developer would expect when reading the code. Creation bytecode is expected to return runtime bytecode, not the result of a sum opperation.

We could, for example, do something like this:

```golang
gas := uint64(10_000)
creationCodePrefix := hexutil.MustDecode("0x600A600C600039600A6000F3") // Return the runtime code
runtimeCode := hexutil.MustDecode("0x607B60005260206000F3")            // Return 123
creationCode := append(creationCodePrefix, runtimeCode...)

t.Run("Create", func(t *testing.T) {
	ret, addr, remainingGas, err := ccEvm.Create(creationCode, gas, new(uint256.Int))
	r.NoError(err)
	r.Equal(runtimeCode, ret)
	r.Equal(runtimeCode, statedb.GetCode(addr))
	r.Equal(crypto.CreateAddress(selfAddr, 0), addr)
	r.Less(remainingGas, gas)
})
t.Run("Create2", func(t *testing.T) {
	salt := new(uint256.Int).SetBytes([]byte("salt"))
	ret, addr, remainingGas, err := ccEvm.Create2(creationCode, gas, new(uint256.Int), salt)
	r.NoError(err)
	r.Equal(runtimeCode, ret)
	r.Equal(runtimeCode, statedb.GetCode(addr))
	r.Equal(crypto.CreateAddress2(selfAddr, salt.Bytes32(), crypto.Keccak256(creationCode)), addr)
	r.Less(remainingGas, gas)
})
```

fabrobles92 committed:
core: apply code review recomendations
